R/bundle.R

```diff
@@ -168,52 +168,101 @@ bundleFiles <- function(appDir) {
+  bundleDir <- bundleAppDir(appDir, appFiles, appPrimaryDoc)
+  on.exit(unlink(bundleDir, recursive = TRUE), add = TRUE)
+
   # generate the manifest and write it into the bundle dir
+  logger("Generate manifest.json")
   manifestJson <- enc2utf8(createAppManifest(bundleDir, appMode,
                                              contentCategory, hasParameters,
                                              appPrimaryDoc,
                                              assetTypeName, users))
   writeLines(manifestJson, file.path(bundleDir, "manifest.json"),
              useBytes = TRUE)
 
-  if (verbose)
-    timestampedLog("Writing Rmd index if necessary")
   # if necessary write an index.htm for shinydoc deployments
+  logger("Writing Rmd index if necessary")
   indexFiles <- writeRmdIndex(appName, bundleDir)
-  on.exit(unlink(indexFiles), add = TRUE)
 
-  if (verbose)
-    timestampedLog("Compressing the bundle")
   # create the bundle and return its path
+  logger("Compressing the bundle")
   prevDir <- setwd(bundleDir)
+
   on.exit(setwd(prevDir), add = TRUE)
   bundlePath <- tempfile("rsconnect-bundle", fileext = ".tar.gz")
   utils::tar(bundlePath, files = ".", compression = "gzip", tar = "internal")
   bundlePath
 }
 
+#' Create a manifest.json describing deployment requirements.
+#'
+#' Given a directory content targeted for deployment, write a manifest.json
+#' into that directory describing the deployment requirements for that
+#' content.
+#'
+#' @param appDir Directory containing the content (Shiny application, R
+#'   Markdown document, etc).
+#'
+#' @param appFiles Optional. The full set of files and directories to be
+#'   included in future deployments of this content. Used when computing
+#'   dependency requirements. When `NULL`, all files in `appDir` are
+#'   considered.
+#'
+#' @param appPrimaryDoc Optional. Specifies the primary document in a content
+#'   directory containing more than one. If `NULL`, the primary document is
+#'   inferred from the file list.
+#'
+#' @param contentCategory Optional. Specifies the kind of content being
+#'   deployed (e.g. `"plot"` or `"site"`).
+#'
+#' @export
+writeManifest <- function(appDir = getwd(),
+                          appFiles = NULL,
+                          appPrimaryDoc = NULL,
+                          contentCategory = NULL) {
+  if (is.null(appFiles)) {
+    appFiles <- bundleFiles(appDir)
+  } else {
+    appFiles <- explodeFiles(appDir, appFiles)
+  }
+
+  appMode <- inferAppMode(appDir, appPrimaryDoc, appFiles)
+  appPrimaryDoc <- inferAppPrimaryDoc(appPrimaryDoc, appFiles, appMode)
+  hasParameters <- appHasParameters(appDir, appPrimaryDoc, appMode, contentCategory)
+
+  # copy files to bundle dir to stage
+  bundleDir <- bundleAppDir(appDir, appFiles, appPrimaryDoc)
+  on.exit(unlink(bundleDir, recursive = TRUE), add = TRUE)
+
+  # generate the manifest and write it into the bundle dir
+  manifestJson <- enc2utf8(createAppManifest(bundleDir,
```

Comment:
For calls to functions that accept a lot of arguments, we generally recommend passing parameters by name rather than positionally, both for clarity and to make future refactoring easier -- we've had a few rsconnect bugs caused by function parameter lists mutating without all the call sites updated.

Reply:
Thanks. This code was essentially duped from `bundleApp`; I'll update calls in both.